packages/@expo/cli/src/start/server/metro/createExpoFallbackResolver.ts

```diff
@@ -0,0 +1,135 @@
+    // both are meant to be installed is direct dependencies, or shouldn't be accessible when
+    // they're fulfilled as isolated dependencies.
+    // The exception are only *optional* peer dependencies, since when they're installed
+    // automatically by newer package manager behaviour, they may become isolated dependencies
+    // that we still wish to access.
+    if (peerDependenciesMeta) {
+      peerDependencies = peerDependencies.filter((dependency) => {
+        const peerMeta = peerDependenciesMeta[dependency];
+        return peerMeta && typeof peerMeta === 'object' && peerMeta.optional === true;
+      });
+    }
+    dependencies.push(...peerDependencies);
+  }
+  // We deduplicate the dependencies and exclude modules that we know are only used for scripts or config-plugins
+  dependencies = dependencies.filter((moduleName, index, dependenciesArr) => {
+    if (EXCLUDE_ORIGIN_MODULES[moduleName]) return false;
+    return dependenciesArr.indexOf(moduleName) === index;
+  });
+  return dependencies.length
+    ? { originModuleName, originModulePath, moduleTestRe: dependenciesToRegex(dependencies) }
+    : null;
+};
+
+/** Creates a fallback module resolver that resolves dependencis of modules named in `originModuleNames` via their path.
+ * @remarks
+ * The fallback resolver targets modules dependended on by modules named in `originModuleNames` and resolves
+ * them from the module root of these origin modules instead.
+ * It should only be used as a fallback after normal Node resolution (and other resolvers) have failed for:
+ * - the `expo` package
+ * - the `expo-router` package
+ * Dependencies mentioned as either optional peer dependencies or direct dependencies by these modules may be isolated
+ * and inaccessible via standard Node module resolution. This may happen when either transpilation adds these
+ * dependencies to other parts of the tree (e.g. `@babel/runtime`) or when a dependency fails to hoist due to either
+ * a corrupted dependency tree or when a peer dependency is fulfilled incorrectly (e.g. `expo-asset`)
+ * @privateRemarks
+ * This does NOT follow Node resolution and is *only* intended to provide a fallback for modules that we depend on
+ * ourselves and know we can resolve (via expo or expo-router)!
+ */
+export function createFallbackModuleResolver({
+  originModuleNames,
+  getStrictResolver,
+}: {
+  originModuleNames: string[];
+  getStrictResolver: StrictResolverFactory;
+}): ExpoCustomMetroResolver {
+  const moduleDescriptions = originModuleNames
+    .map(getModuleDescription)
+    .filter(
+      (moduleDescription): moduleDescription is ModuleDescription => moduleDescription != null
+    );
+
+  return function requestFallbackModule(immutableContext, moduleName, platform) {
+    // Early return if `moduleName` cannot be a module specifier
+    if (moduleName[0] === '.' || moduleName[0] === '/') {
+      return null;
+    }
```

Comment:
Does this also need to account for alias tokens? E.g. `@/module` or `~/module`?

Reply:
not necessarily. The module resolver (almost always) triggers when module resolution has otherwise failed. But this is just to avoid the regex checks as often as possible for imports that can never be module specifiers. It's ok if we miss some (and things like the tsconfig resolver come first anyway), since the regexes are what match the module specifiers we want to target